src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,881 @@
+        logger.log(Level.INFO, String.format("Found %d exons.", exonDB.exonCount()));
+        logger.log(Level.INFO, "Reading exons locations from intervals done.");
+
+        // Initializing count and count column management member fields:
+        countColumns = groupBy.countColumns(this);
+        final int columnCount = countColumns.columnCount();
+        counts = new int[columnCount][exonDB.exonCount()];
+        columnTotals = new long[columnCount];
+
+        // Open output files and write headers:
+        outputWriter = openOutputWriter(output,composeMatrixOutputHeader(getCommandLine(),groupBy,countColumns.columnNames()));
+        if (columnSummaryOutput != null) {
+            columnSummaryOutputWriter = openOutputWriter(columnSummaryOutput,
+                    composeColumnSummaryHeader(getCommandLine(),groupBy,exonDB.exonCount(),exonDB.exomeSize()));
+        }
+        if (rowSummaryOutput != null) {
+            rowSummaryOutputWriter = openOutputWriter(rowSummaryOutput,
+                    composeRowOutputHeader(getCommandLine(),groupBy,countColumns.columnCount()));
+        }
+
+        // Next we start the traversal:
+        logger.log(Level.INFO,"Collecting read counts ...");
+    }
+
+    /**
+     * Opens the output file for writing with a print-writer.
+     * @param output the output file.
+     * @param headerText to be printed immediately after opening the writer.
+     *
+     * @throws UserException.CouldNotCreateOutputFile if there was some problem creating or overwriting {@code output}.
+     * @return never {@code null}.
+     */
+    private PrintWriter openOutputWriter(final File output, final String headerText) {
+        try {
+            final PrintWriter result = new PrintWriter(output);
+            result.println(headerText);
+            result.flush();
+            return result;
+        } catch (final IOException e) {
+            throw new UserException.CouldNotCreateOutputFile(output,e);
+        }
+    }
+
+    @Override
+    public void apply(final SAMRecord read, final ReferenceContext referenceContext, final FeatureContext featureContext) {
+        final GenomeLoc readLocation = genomeLocFactory.createGenomeLoc(read);
+
+        final int columnIndex = countColumns.columnIndex(read);
+        if (columnIndex >= 0) { // < 0 would mean means that the read is to be ignored.
+            exonDB.indexRange(readLocation).forEach(i -> {
+                counts[columnIndex][i]++;
+                columnTotals[columnIndex]++;
+            });
+        }
+    }
+
+    @Override
+    public Object onTraversalDone() {
+        super.onTraversalDone();
+        logger.log(Level.INFO, "Collecting read counts done.");
+
+        logger.log(Level.INFO, "Writing counts ...");
+        final int exonCount = exonDB.exonCount();
+        final int columnCount = counts.length;
+        final int countColumnCount = countColumns.columnCount();
+        final int[] countBuffer = new int[countColumnCount];
```

Comment:
this temp only lives inside the `i` loop. Move it inside to limit confusion of the reader.